runtime/src/validate_block/storage_functions.rs

```diff
@@ -0,0 +1,79 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Cumulus.  If not, see <http://www.gnu.org/licenses/>.
+
+//! All storage functions that are replaced by `validate_block` in the Substrate runtime.
+
+use crate::WitnessData;
+use rstd::{slice, ptr, cmp};
+
+pub static mut STORAGE: Option<WitnessData> = None;
+const STORAGE_SET_EXPECT: &str = "`STORAGE` needs to be set before calling this function.";
+
+pub unsafe fn ext_get_allocated_storage(key_data: *const u8, key_len: u32, written_out: *mut u32) -> *mut u8 {
```

Comment:
The signature of this function is essentially `get_allocated_storage(key) -> &mut [u8]`.

It can be called twice in a row, which would give aliased mutable memory -- we should make sure the same invariants of the substrate-native implementation are preserved.

Reply:
this and all other functions in this module should have clear documentation about invariants and restrictions on arguments -- we generally have a "all public items should have docs" guideline but this is especially true for `unsafe` code.